### PR TITLE
Adding dry-run feature and potential new way of testing

### PR DESCRIPTION
**Pull request by swenzel**
Hey there,

I've implemented the dry-run feature and in order to test it I came up with the following test procedure:

1. create a testing folder in the temporary directory (`<tmp>/bump-test/`).
2. create an empty bare repository (`<tmp>/bump-test/sampleRepo.git`).
3. clone it into `<tmp>/bump-test/sampleRepo`
4. copy contents from `grunt-bump/test/sampleRepo`
5. copy grunt, semver and the essential parts of grunt-bump into `<tmp>/bump-test/sampleRepo/node_modules`
6. test grunt-bump using execSync to run `git ...` and `grunt ...` as child processes within `<tmp>/bump-test/sampleRepo`
7. delete `<tmp>/bump-test` with all its subfolders and -directories

Currently there are some issues though.
Once #105 is merged, I might change the tests to use mocha instead of node unit... wonder how long that will take though :/

Greetings
Swen

### CODE REVIEW

**Comment by eddiemonge** [comment]
i think it would probably be better to mock out git for tests instead of jumping through all those hoops to make it work

File: README.md [hunk 1]
```diff
@@ -188,6 +188,34 @@ $ grunt changelog
...
+$ grunt bump --dry-run
+Running "bump" task
+Running grunt-bump in dry mode!
+>> bump-dry: Version bumped to 1.0.1 (in package.json)
+>> bump-dry: git commit package.json -m "Release v1.0.1"
+>> bump-dry: git tag -a v1.0.1 -m "Version 1.0.1"
+>> bump-dry: git push origin && git push origin --tags
+```
+
+Since the tag is parsed and forwarded by grunt, it will also be awailable if you
```
Comment (eddiemonge) [inline README.md]: typo in `available`

File: README.md [hunk 2]
```diff
@@ -188,6 +188,34 @@ $ grunt changelog
...
+call bump within other tasks:
```
Comment (eddiemonge) [inline README.md]: i think if its clear that you can pass `--dry-run` to grunt then this example is redundant

File: package.json [hunk 3]
```diff
@@ -28,7 +28,8 @@
     "Snugug <snugug@gmail.com>",
     "angleman <angleman@users.noreply.github.com>",
     "cattail <zhongchiyu@gmail.com>",
-    "ylesaout <ylesaout@gmail.com>"
+    "ylesaout <ylesaout@gmail.com>",
+    "Swen Wenzel <swen.wenzel@t-online.de>"
```
Comment (eddiemonge) [inline package.json]: please remove this as it will automatically be appended

File: tasks/bump.js [hunk 4]
```diff
@@ -20,6 +20,8 @@ var exec = require('child_process').exec;
 module.exports = function(grunt) {
 
   var DESC = 'Increment the version, commit, tag and push.';
+  var DRY = false;
```
Comment (eddiemonge) [inline tasks/bump.js]: `var DRY = grunt.option('dry-run')`

File: tasks/bump.js [hunk 5]
```diff
@@ -37,23 +39,31 @@ module.exports = function(grunt) {
       globalReplace: false
     });
 
+    if (grunt.option('dry-run')) {
```
Comment (eddiemonge) [inline tasks/bump.js]:
```javascript
if (DRY) {
  grunt.log....
```

File: tasks/bump.js [hunk 6]
```diff
@@ -37,23 +39,31 @@ module.exports = function(grunt) {
...
       opts.push = false;
-    }
-
-    if (incOrCommitOnly === 'commit-only') {
+    } else if (incOrCommitOnly === 'commit-only') {
       grunt.verbose.writeln('Only committing/tagging/pushing.');
-
       opts.bumpVersion = false;
+    } else if (incOrCommitOnly !== undefined && incOrCommitOnly !== '') {
+      grunt.fail.fatal('Invalid value "' + incOrCommitOnly + '" for 2nd argument!');
```
Comment (eddiemonge) [inline tasks/bump.js]: should these lines be in a separate commit?

File: tasks/bump.js [hunk 7]
```diff
@@ -37,23 +39,31 @@ module.exports = function(grunt) {
...
     }
 
     var exactVersionToSet = grunt.option('setversion');
     if (!semver.valid(exactVersionToSet)) {
-        exactVersionToSet = false;
+      // should not silently fail, the user DID try to use a tag.
```
Comment (eddiemonge) [inline tasks/bump.js]: separate commit

**Comment by eddiemonge** [comment]
please separate the commits out into separate pull requests. I like the dry run stuff but want it mergeable individually

**Comment by swenzel** [comment]
Okay no problem, how about the testing? Shall I leave it like this for now?

File: tasks/bump.js [hunk 8]
```diff
@@ -107,8 +112,12 @@ module.exports = function(grunt) {
           grunt.fatal('Can not find a version to bump in ' + file);
         }
 
-        grunt.file.write(file, content);
-        grunt.log.ok('Version bumped to ' + version + (opts.files.length > 1 ? ' (in ' + file + ')' : ''));
+        if (!DRY){
+          grunt.file.write(file, content);
+          grunt.log.ok('Version bumped to ' + version +  ' (in ' + file + ')');
```
Comment (eddiemonge) [inline tasks/bump.js]: move this msg to a variable and pass it in here and then concat it in line 119

File: tasks/bump.js [hunk 9]
```diff
@@ -149,14 +158,22 @@ module.exports = function(grunt) {
...
 
-      exec('git commit ' + opts.commitFiles.join(' ') + ' -m "' + commitMessage + '"', function(err, stdout, stderr) {
-        if (err) {
-          grunt.fatal('Can not create the commit:\n  ' + stderr);
-        }
-        grunt.log.ok('Committed as "' + commitMessage + '"');
+      if(DRY){
+        // since nothing has actually changed, running git commit with --dry-run
+        // will fail, so we just print the command instead
+        grunt.log.ok('bump-dry: '+cmd);
```
Comment (eddiemonge) [inline tasks/bump.js]: spaces around the plus sign

File: tasks/bump.js [hunk 10]
```diff
@@ -165,25 +182,40 @@ module.exports = function(grunt) {
...
 
-      exec('git tag -a ' + tagName + ' -m "' + tagMessage + '"' , function(err, stdout, stderr) {
-        if (err) {
-          grunt.fatal('Can not create the tag:\n  ' + stderr);
-        }
-        grunt.log.ok('Tagged as "' + tagName + '"');
+      var cmd = 'git tag -a ' + tagName + ' -m "' + tagMessage + '"';
+      if (DRY) {
+        // cannot run git tag in dry mode so we just give the command as output.
+        grunt.log.ok('bump-dry: '+cmd);
```
Comment (eddiemonge) [inline tasks/bump.js]: spaces around the plus sign

File: tasks/bump.js [hunk 11]
```diff
@@ -165,25 +182,40 @@ module.exports = function(grunt) {
...
-      exec('git push ' + opts.pushTo + ' && git push ' + opts.pushTo + ' --tags', function(err, stdout, stderr) {
-        if (err) {
-          grunt.fatal('Can not push to ' + opts.pushTo + ':\n  ' + stderr);
-        }
-        grunt.log.ok('Pushed to ' + opts.pushTo);
+      var cmd = 'git push ' + opts.pushTo + ' && git push ' + opts.pushTo + ' --tags';
+      if (DRY){
+        // since nothing has actually been comitted, running git push with --dry-run
+        // will fail, so we just print the command instead
+        grunt.log.ok('bump-dry: '+cmd);
```
Comment (eddiemonge) [inline tasks/bump.js]: spaces around the plus sign

**Comment by eddiemonge** [comment]
without the tests in this one are fine for now. a few more comments.

thanks for the work on this

**Comment by swenzel** [comment]
You're not easy to please xD
What kind of comments do you want? I thought due to the naming the parts were pretty self explanatory...

**Comment by eddiemonge** [comment]
I meant my comments on the pull request itself :)  the code doesn't need commenting.

**Comment by swenzel** [comment]
Ah okay :)
Here you go!

File: tasks/bump.js [hunk 12]
```diff
@@ -107,8 +112,13 @@ module.exports = function(grunt) {
           grunt.fatal('Can not find a version to bump in ' + file);
         }
 
-        grunt.file.write(file, content);
-        grunt.log.ok('Version bumped to ' + version + (opts.files.length > 1 ? ' (in ' + file + ')' : ''));
+        var logMsg = 'Version bumped to ' + version +  ' (in ' + file + ')';
```
Comment (eddiemonge) [inline tasks/bump.js]: sneaky change here in the output. I guess its fine though

File: tasks/bump.js [hunk 13]
```diff
@@ -165,25 +183,40 @@ module.exports = function(grunt) {
...
     // PUSH CHANGES
     runIf(opts.push, function() {
-      exec('git push ' + opts.pushTo + ' && git push ' + opts.pushTo + ' --tags', function(err, stdout, stderr) {
-        if (err) {
-          grunt.fatal('Can not push to ' + opts.pushTo + ':\n  ' + stderr);
-        }
-        grunt.log.ok('Pushed to ' + opts.pushTo);
+      var cmd = 'git push ' + opts.pushTo + ' && git push ' + opts.pushTo + ' --tags';
+      if (DRY){
+        // since nothing has actually been comitted, running git push with --dry-run
```
Comment (eddiemonge) [inline tasks/bump.js]: these comments arent really needed

**Comment by eddiemonge** [comment]
merged in https://github.com/vojtajina/grunt-bump/commit/10f5ecdab375f6eb68e4d750be768f5f91208dea
